### DISCUSSION TITLE
project.protectedbranches.delete 404: 404 Not found

### DISCUSSION THREAD

**Discussion by jalanh11640** [python-gitlab/python-gitlab#2536] [Q&A] [answered]
I have a script I've written that creates/configures a gitlab project.  After creating the project, the script does a "project.protectedbranches.delete('work')".  'work' is our default branch so it will always exist and is a protected branch.  About half the time I make this call, I get the following callback.  Is there some server/client timing that I need to be aware of?  If so, is there a way to query to know that it is safe to delete the protections?

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/gitlab/exceptions.py", line 337, in wrapped_f
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/gitlab/mixins.py", line 436, in delete
    self.gitlab.http_delete(path, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/gitlab/client.py", line 1102, in http_delete
    return self.http_request("delete", path, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/gitlab/client.py", line 803, in http_request
    response_body=result.content,
gitlab.exceptions.GitlabHttpError: 404: 404 Not found

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./create_project", line 190, in <module>
    main()
  File "./create_project", line 185, in main
    configure_branches(project, not arguments.maintainer, arguments.nobranches)
  File "./create_project", line 162, in configure_branches
    project.protectedbranches.delete('work')
  File "/usr/local/lib/python3.7/site-packages/gitlab/exceptions.py", line 339, in wrapped_f
    raise error(e.error_message, e.response_code, e.response_body) from e
gitlab.exceptions.GitlabDeleteError: 404: 404 Not found

python: 3.7.9
python-gitlab version: 3.12.0
gitlab version: GitLab Enterprise Edition 15.5.9-ee

**Comment by nejch** [comment, marked as answer]
Hey @jalanh11640! It's hard to say without seeing your code.

But yes, GitLab does a lot of things asynchronously these days, which can take some time especially with things like creating the initial repository, branches, merge requests etc. Try with a simple sleep first, maybe a few seconds, and if that works you can then do it in a smarter loop until success or so.

There's no harm in just putting this into try/except blocks until the resource is available to make requests against. It'll be fewer API calls than trying to query the state of the project.

**Reply by jalanh11640** [reply]
Thanks, that's not a particularly satisfying answer, but I appreciate the answer.

I've updated my code to do this.. and it seems to work fine.

    # Sometimes the delete results in an exception.  Keep trying until successful
    tryDelete = True
    while tryDelete:
        try:
            # 'delete' the work branch,this doesn't delete it, just unprotects it.
            project.protectedbranches.delete('work')
            tryDelete = False
        except gitlab.exceptions.GitlabDeleteError:
            pass

Thanks

**Reply by nejch** [reply]
@jalanh11640 there is nothing python-gitlab can do about this - you'd hit this issue with any other API client or curl as well. It may be better to open an issue upstream on GitLab if you think the error messages returned from the server are not informative enough, we really just make requests against the server.

Glad you got it to work!
